src/coreclr/jit/morph.cpp

```diff
@@ -13446,6 +13441,26 @@ GenTree* Compiler::fgOptimizeEqualityComparisonWithConst(GenTreeOp* cmp)
     {
         ssize_t op2Value = static_cast<ssize_t>(op2->IntegralValue());
 
+        // Optimizes (X & 1) == 1 to (X & 1)
```

Comment:
This also needs to check that we will be returning a properly typed tree here. `X & 1` can be of `TYP_LONG`, while relops are always `TYP_INT`.

There are two ways to fix this: a) just give up if the `AND` is not of `TYP_INT` (`genActualType(andNode) == relop->TypeGet()`), or try our luck with `optNarrowTree` (see example of usage in this function). Note that `optNarrowTree` should only be called under an `fgGlobalMorph` guard.

The `optNarrowTree` solution should get better diffs, the "always check types" one is simpler, up to you which one to implement.

Fixing this should fix the x86 failures we're seeing in CI.

(Also, for debugging, [SPMI](https://github.com/dotnet/runtime/blob/main/src/coreclr/scripts/superpmi.md) is a very useful tool.)